src/proxyp/Parser.cc

```diff
@@ -0,0 +1,287 @@
+
+    extractPort(tok, message->sourceAddress, true);
+    extractPort(tok, message->destinationAddress, false);
+}
+
+/// parses PROXY protocol v1 message from the buffer
+ProxyProtocol::Parsed
+ProxyProtocol::One::Parse(const SBuf &buf)
+{
+    Parser::Tokenizer tok(buf);
+
+    static const SBuf::size_type maxMessageLength = 107; // including CRLF
+    static const auto maxInteriorLength = maxMessageLength - Magic.length() - 2;
+    static const auto interiorChars = CharacterSet::CR.complement().rename("non-CR");
+    SBuf interior;
+
+    if (!(tok.prefix(interior, interiorChars, maxInteriorLength) &&
+            tok.skip('\r') &&
+            tok.skip('\n'))) {
+        if (tok.atEnd())
+            throw Parser::BinaryTokenizer::InsufficientInput();
+        else if (interior.isEmpty())
+            throw TexcHere("PROXY/1.0 error: the message block is empty");
+        else
+            throw TexcHere("PROXY/1.0 error: missing CRLF in the message");
+    }
+    // grabbed all header bytes
+
+    MessagePointer message = new Message("1.0");
+
+    Parser::Tokenizer interiorTok(interior);
+
+    if (!interiorTok.skip(' '))
+        throw TexcHere("PROXY/1.0 error: missing SP after the magic sequence");
+
+    static const SBuf protoUnknown("UNKNOWN");
+    static const SBuf protoTcp("TCP");
+
+    if (interiorTok.skip(protoTcp))
+        parseAddresses(interiorTok, message);
+    else if (interiorTok.skip(protoUnknown))
+        message->ignoreAddresses();
+    else
+        throw TexcHere("PROXY/1.0 error: invalid INET protocol or family");
+
+    return Parsed(message, tok.parsedSize());
+}
+
+void
+ProxyProtocol::Two::parseAddresses(const uint8_t family, Parser::BinaryTokenizer &tok, MessagePointer &message)
+{
+    switch (family) {
+
+    case afInet: {
+        message->sourceAddress = tok.inet4("src_addr IPv4");
+        message->destinationAddress = tok.inet4("dst_addr IPv4");
+        message->sourceAddress.port(tok.uint16("src_port"));
+        message->destinationAddress.port(tok.uint16("dst_port"));
+        break;
+    }
+
+    case afInet6: {
+        message->sourceAddress = tok.inet6("src_addr IPv6");
+        message->destinationAddress = tok.inet6("dst_addr IPv6");
+        message->sourceAddress.port(tok.uint16("src_port"));
+        message->destinationAddress.port(tok.uint16("dst_port"));
+        break;
+    }
+
+    case afUnix: { // TODO: add support
+        // the address block length is 216 bytes
+        tok.skip(216, "unix_addr");
+        break;
+    }
+
+    default: {
+        // unreachable code: we have checked family validity already
+        Must(false);
+        break;
+    }
+    }
+}
+
+void
+ProxyProtocol::Two::parseTLVs(Parser::BinaryTokenizer &tok, MessagePointer &message) {
+    while (!tok.atEnd()) {
+        const auto type = tok.uint8("pp2_tlv::type");
+        message->tlvs.emplace_back(type, tok.pstring16("pp2_tlv length and value"));
```

Comment:
```suggestion
        message->tlvs.emplace_back(type, tok.pstring16("pp2_tlv::value"));
```

pstring16() adds `.length` when parsing the length part of the string.